pkg/daemon/ceph/client/rados.go

```diff
@@ -0,0 +1,172 @@
+	Tag     string            `json:"tag"`
+	Lockers []radosLockerInfo `json:"lockers"`
+}
+
+type radosLockerInfo struct {
+	Name   string `json:"name"`
+	Cookie string `json:"cookie"`
+}
+
+// RadosLockObject locks a rados object in a given pool and namespace and returns a lock "cookie"
+// that can be used to identify this unique lock.
+func RadosLockObject(
+	context *clusterd.Context, clusterInfo *ClusterInfo,
+	pool, namespace, objectName, lockName string,
+	lockTimeout time.Duration,
+) (string, error) {
+	// generate a random "cookie" that identifies this lock "session"
+	cookieLen := 12
+	b := make([]byte, cookieLen)
+	rand.Read(b)
+	cookie := fmt.Sprintf("%x", b)[:cookieLen]
+
+	cmd := NewRadosCommand(context, clusterInfo, []string{
+		"--pool", pool,
+		"--namespace", namespace,
+		"lock", "get", objectName, lockName,
+		"--lock-tag", lockName, // assume we aren't making many locks; use lock name as the tag
+		"--lock-cookie", cookie,
+		"--lock-duration", fmt.Sprintf("%d", int(math.Ceil(lockTimeout.Seconds()))),
+	})
+	if _, err := cmd.RunWithTimeout(exec.CephCommandsTimeout); err != nil {
+		return "", errors.Wrapf(err, "failed to aquire lock %q on rados object rados://%s/%s/%s",
+			lockName, pool, namespace, objectName)
+	}
+
+	return cookie, nil
+}
+
+// RadosUnlockObject unlocks a rados object in a given pool and namespace by searching for locks on
+// the object that match the "cookie" obtained from a RadosLockObject() call.
+func RadosUnlockObject(
+	context *clusterd.Context, clusterInfo *ClusterInfo,
+	pool, namespace, objectName, lockName string,
+	lockCookie string,
+) error {
+	lockInfo, err := radosObjectLockInfo(context, clusterInfo, pool, namespace, objectName, lockName, lockCookie)
+	if err != nil {
+		return errors.Wrap(err, "failed to unlock object")
+	}
+
+	if lockInfo.Name != lockName || lockInfo.Tag != lockName {
+		// some other lock has locked the object, but that means the lock with the given cookie no
+		// longer has it locked; treat this as success
+		logger.Infof("rados object rados://%s/%s/%s is not locked by lock %q but is locked by %q",
+			pool, namespace, objectName, lockName, lockInfo.Name)
+		return nil
+	}
+
+	if len(lockInfo.Lockers) == 0 {
+		logger.Infof("rados object rados://%s/%s/%s is already fully unlocked (lock: %q, cookie: %q)",
+			pool, namespace, objectName, lockName, lockCookie)
+		return nil
+	}
+
+	locker := findLockerWithCookie(lockInfo.Lockers, lockCookie)
+	if locker == nil {
+		// this cookie is not locking the object but another is; still treat this as success but log it
+		logger.Infof("rados object rados://%s/%s/%s is not locked by lock %q with cookie %q but is locked: %+v",
+			pool, namespace, objectName, lockName, lockCookie, lockInfo.Lockers)
+		return nil
+	}
+
+	cmd := NewRadosCommand(context, clusterInfo, []string{
+		"--pool", pool,
+		"--namespace", namespace,
+		"lock", "break", objectName, lockName, locker.Name, // breaking the lock also requires locker name
```

Comment:
Ultimately, the `lock break` command requires the locker name, which is why this function is so long.